File: backend/app/schemas/feedback.py
from __future__ import annotations

from pydantic import BaseModel


class FeedbackRequest(BaseModel):
    trace_id: str
    user_action: str  # "selected", "rerolled", "rejected", "voice_escape"
    selection_speed_ms: int | None = None
    selected_index: int | None = None
    reroll_count: int | None = None
    plan_approved: bool | None = None
    plan_approval_button: str | None = None
    used_unhinged_modifier: bool | None = None


class FeedbackResponse(BaseModel):
    status: str
    computed_reward: float | None = None

File: src/lib/prediction/feedback.ts
// ─── Feedback Module ─────────────────────────────────────────────────────────
// Fire-and-forget feedback to the FastAPI backend for RLHF scoring via Langfuse.
// Silently drops if no auth token or no trace_id (mock mode).

import { getAccessToken, getBackendUrl } from '../auth/auth';

export interface FeedbackPayload {
  trace_id: string;
  user_action: 'selected' | 'rerolled' | 'rejected';
  selection_speed_ms?: number;
  selected_index?: number;
  reroll_count?: number;
  plan_approved?: boolean;
  plan_approval_button?: string;
  used_unhinged_modifier?: boolean;
}

/**
 * Send feedback to backend. Fire-and-forget — no await needed.
 * Returns immediately if no auth token or no trace_id.
 */
export function sendFeedback(payload: FeedbackPayload): void {
  const token = getAccessToken();
  if (!token || !payload.trace_id) return;

  const backendUrl = getBackendUrl();

  fetch(`${backendUrl}/api/v1/feedback`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${token}`,
    },
    body: JSON.stringify(payload),
  }).catch(e => {
    console.warn('Feedback send failed:', e);
  });
}
